rust/src/operations/mod.rs

```diff
@@ -0,0 +1,285 @@
+
+    /// Error returned when the table to be created already exists
+    #[error("Table: '{0}' already exists")]
+    TableAlreadyExists(String),
+
+    /// Error returned when errors occur in underlying delta table instance
+    #[error("Error in underlying DeltaTable")]
+    DeltaTableError {
+        /// Raw internal DeltaTableError
+        #[from]
+        source: DeltaTableError,
+    },
+
+    /// Errors occurring inside the DeltaWriter modules
+    #[error("Error in underlying storage backend")]
+    DeltaWriterError {
+        /// Raw internal StorageError
+        #[from]
+        source: DeltaWriterError,
+    },
+
+    /// Error returned when errors occur in underlying storage instance
+    #[error("Error in underlying storage backend")]
+    StorageError {
+        /// Raw internal StorageError
+        #[from]
+        source: StorageError,
+    },
+
+    /// Error returned when errors occur in underlying storage instance
+    #[error("Error handling arrow data")]
+    ArrowError {
+        /// Raw internal StorageError
+        #[from]
+        source: ArrowError,
+    },
+
+    /// Error returned for errors internal to Datafusion
+    #[error("Error handling arrow data")]
+    DataFusionError {
+        /// Raw internal DataFusionError
+        #[from]
+        source: DataFusionError,
+    },
+}
+
+fn to_datafusion_err(e: impl std::error::Error) -> DataFusionError {
+    DataFusionError::Plan(e.to_string())
+}
+
+/// High level interface for executing commands against a DeltaTable
+pub struct DeltaCommands {
+    table: DeltaTable,
+}
+
+impl DeltaCommands {
+    /// load table from uri
+    pub async fn try_from_uri<T>(uri: T) -> DeltaCommandResult<Self>
+    where
+        T: Into<String>,
+    {
+        let table = get_table_from_uri_without_update(uri.into())?;
+        Ok(Self { table })
+    }
+
+    /// Get a reference to the underlying table
+    pub fn table(&self) -> &DeltaTable {
+        &self.table
+    }
+
+    async fn execute(
+        &mut self,
+        operation: DeltaOperation,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> DeltaCommandResult<()> {
+        let transaction = Arc::new(DeltaTransactionPlan::new(
+            self.table.table_uri.clone(),
+            plan,
+            operation,
+            None,
+        ));
+
+        let session_ctx = SessionContext::new();
+        let task_ctx = session_ctx.task_ctx();
+        let _ = collect(transaction, task_ctx).await?;
+        self.table.update().await?;
+
+        Ok(())
+    }
+
+    /// Create a new Delta table
+    pub async fn create(
+        &mut self,
+        metadata: DeltaTableMetaData,
+        mode: Option<SaveMode>,
+    ) -> DeltaCommandResult<()> {
+        let operation = DeltaOperation::Create {
+            mode: mode.clone().unwrap_or(SaveMode::Ignore),
+            metadata: metadata.clone(),
+            location: self.table.table_uri.clone(),
+            // TODO get the protocol from somewhere central
```

Comment:
Yeah if I understand correctly, the protocol should be determined based on the existing protocol version (if any) and the latest metadata (which should contain all the information to determine the required protocol for the requested table schema, config, etc.).